wrappers/dbus.go

```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		return "", err
+	}
+	snapKey := []byte("\nX-Snap=")
+	pos := bytes.Index(content, snapKey)
+	if pos == -1 {
+		return "", nil
+	}
+	snapName := content[pos+len(snapKey):]
+	pos = bytes.IndexRune(snapName, '\n')
+	if pos != -1 {
+		snapName = snapName[:pos]
+	}
+	return string(snapName), nil
+}
+
+// snapServiceActivationFiles returns the list of service activation files for a snap.
+func snapServiceActivationFiles(dir, snapName string) (services []string, err error) {
+	glob := filepath.Join(dir, "*.service")
+	matches, err := filepath.Glob(glob)
+	if err != nil {
+		return nil, err
+	}
+	for _, match := range matches {
+		serviceSnap, err := snapNameFromServiceFile(match)
+		if err != nil {
+			return nil, err
+		}
+		if serviceSnap == snapName {
+			services = append(services, filepath.Base(match))
+		}
+	}
+	return services, nil
+}
+
+func AddSnapDBusActivationFiles(s *snap.Info) error {
+	if err := os.MkdirAll(dirs.SnapDBusSessionServicesDir, 0755); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(dirs.SnapDBusSystemServicesDir, 0755); err != nil {
+		return err
+	}
+
+	// Make sure we include any service files that claim to have
+	// been written by the snap.
+	sessionServices, err := snapServiceActivationFiles(dirs.SnapDBusSessionServicesDir, s.InstanceName())
+	if err != nil {
+		return err
+	}
+	systemServices, err := snapServiceActivationFiles(dirs.SnapDBusSystemServicesDir, s.InstanceName())
+	if err != nil {
+		return err
+	}
+
+	sessionContent := make(map[string]osutil.FileState)
+	systemContent := make(map[string]osutil.FileState)
+
+	for _, app := range s.Apps {
+		if !app.IsService() {
+			continue
+		}
+
+		for _, slot := range app.ActivatesOn {
+			var busName string
+			if err := slot.Attr("name", &busName); err != nil {
+				return err
+			}
+
+			content, err := generateDBusActivationFile(app, busName)
+			if err != nil {
+				return err
+			}
+			filename := busName + ".service"
+			fileState := &osutil.MemoryFileState{
+				Content: content,
+				Mode:    0644,
+			}
+			switch app.DaemonScope {
+			case snap.SystemDaemon:
+				systemContent[filename] = fileState
+				systemServices = append(systemServices, filename)
+			case snap.UserDaemon:
+				sessionContent[filename] = fileState
+				sessionServices = append(sessionServices, filename)
+			}
+		}
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapDBusSessionServicesDir, sessionServices, sessionContent)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapDBusSystemServicesDir, systemServices, systemContent)
+	if err != nil {
+		// On error, remove files installed by first invocation
+		osutil.EnsureDirStateGlobs(dirs.SnapDBusSessionServicesDir, sessionServices, nil)
+	}
+	return err
+}
+
+func RemoveSnapDBusActivationFiles(s *snap.Info) error {
+	// Select files to delete via X-Snap line to ensure everything
+	// is cleaned up if "snap try" is used and snap.yaml is
+	// modified.
```

Comment:
Thanks for this comment, I was naively thinking this could be optimized by checking if we have any .ActivatesOn slots but yes, not way around it. I wonder if we could have a trivial snap try spread test, but maybe it's a material for a followup at this point.